parsing/tools/src/main/java/com/apache/calcite/parsing/tools/FindDialectParsingErrors.java

```diff
@@ -0,0 +1,424 @@
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Processes a CSV file containing queries and will output a JSON file
+ * containing the results of failing queries.
+ */
+public class FindDialectParsingErrors {
+
+  private final String inputPath;
+  private final String outputPath;
+  private final Dialect dialect;
+  private final boolean groupByErrors;
+  private final int numSampleQueries;
+  private final List<MessageFormat> errorFormats;
+
+  /**
+   *  Creates a new instance of {@code FindDialectParsingErrors}. Also populates
+   *  the errorFormats list with a {@code MessageFormat} object for each custom
+   *  defined error message in CalciteResource.
+   *
+   * @param inputPath Path to the input CSV file containing queries
+   * @param outputPath Path to the output JSON file containing results of
+   *                   failing queries
+   * @param dialect Specifies which dialectic parser to use for processing
+   *                queries
+   * @param groupByErrors Specifies if the output format should group failing
+   *                      queries by error message type
+   * @param numSampleQueries The max number of sample queries to show for an
+   *                         error type when grouping queries by error message
+   *                         type
+   */
+  public FindDialectParsingErrors(String inputPath, String outputPath,
+      Dialect dialect, boolean groupByErrors, int numSampleQueries) {
+    this.inputPath = inputPath;
+    this.outputPath = outputPath;
+    this.dialect = dialect;
+    this.groupByErrors = groupByErrors;
+    this.numSampleQueries = numSampleQueries;
+    errorFormats = new ArrayList<>();
+    Method[] methods = CalciteResource.class.getMethods();
+    for (Method m : methods) {
+      errorFormats.add(
+          new MessageFormat(
+              m.getAnnotationsByType(
+              Resources.BaseMessage.class)[0].value(), Locale.ROOT)
+      );
+    }
+  }
+
+  /**
+   * Runs one of the processing methods and will output a JSON file containing
+   * the results of failing queries. This will run findErrorGroups if
+   * groupByErrors is true. Otherwise, findFullErrors will run.
+   *
+   * @throws IOException If it fails to read the input file
+   */
+  public void run() throws IOException {
+    CSVReader reader = new CSVReader(
+        new InputStreamReader(new FileInputStream(inputPath),
+        StandardCharsets.UTF_8), ',', '"', 1);
+    List<String[]> rows = reader.readAll();
+    reader.close();
+    List<String> queries = rows.parallelStream()
+        .map(row -> sanitize(row[0]))
+        .collect(Collectors.toList());
+    if (groupByErrors) {
+      findErrorGroups(queries);
+    } else {
+      findFullErrors(queries);
+    }
+  }
+
+  /**
+   * This is the default processing method. It parallel processes the list of
```

Comment:
nit: It processes the list of queries in parallel....

Reply:
Updated